runtime/src/accounts_db.rs

```diff
@@ -713,26 +723,35 @@ impl AccountsDB {
         // Calculate store counts as if everything was purged
         // Then purge if we can
         let mut store_counts: HashMap<AppendVecId, (usize, HashSet<Pubkey>)> = HashMap::new();
-        let storage = self.storage.read().unwrap();
-        for (key, (account_infos, _ref_count)) in &purges {
-            for (slot, account_info) in account_infos {
-                let slot_storage = storage.0.get(&slot).unwrap();
-                let store = slot_storage.get(&account_info.store_id).unwrap();
+        for (key, (account_infos, ref_count)) in purges.iter_mut() {
+            account_infos.retain(|(slot, account_info)| {
+                if old_rooted_dead_slots.contains(slot) {
+                    *ref_count = self
+                        .accounts_index
+                        .read()
+                        .unwrap()
+                        .ref_count_from_storage(&key);
+                    return false;
+                }
                 if let Some(store_count) = store_counts.get_mut(&account_info.store_id) {
                     store_count.0 -= 1;
```

Comment:
This and the matching `count - 1` below should only be decremented if the `clean_old_rooted_accounts` above didn't clean those storage entries already

Reply:
Fixed and updated in latest commits.